client/chain-spec/src/chain_spec.rs

```diff
@@ -321,47 +549,73 @@ impl<G, E: serde::de::DeserializeOwned> ChainSpec<G, E> {
+
+		Ok(ChainSpec { client_spec, genesis: GenesisSource::File(path), code: code.code })
 	}
 }
 
+/// Helper structure for serializing (and only serializing) the ChainSpec into JSON file. It
+/// represents the layout of `ChainSpec` JSON file.
 #[derive(Serialize, Deserialize)]
-struct JsonContainer<G, E> {
+#[serde(deny_unknown_fields)]
+struct ChainSpecJsonContainer<G, E> {
 	#[serde(flatten)]
 	client_spec: ClientSpec<E>,
 	genesis: Genesis<G>,
+	#[serde(with = "sp_core::bytes", skip_serializing_if = "Vec::is_empty")]
+	code: Vec<u8>,
 }
 
 impl<G: RuntimeGenesis, E: serde::Serialize + Clone + 'static> ChainSpec<G, E> {
-	fn json_container(&self, raw: bool) -> Result<JsonContainer<G, E>, String> {
-		let genesis = match (raw, self.genesis.resolve()?) {
+	fn json_container(&self, raw: bool) -> Result<ChainSpecJsonContainer<G, E>, String> {
+		let mut raw_genesis = match (raw, self.genesis.resolve()?) {
+			(true, Genesis::RuntimeGenesisConfigPatch(patch)) => {
+				let storage = RuntimeCaller::new(&self.code[..]).get_storage_for_patch(patch)?;
+				RawGenesis::from(storage)
+			},
+			(true, Genesis::RuntimeGenesisConfig(config)) => {
+				let storage = RuntimeCaller::new(&self.code[..]).get_storage_for_config(config)?;
+				RawGenesis::from(storage)
+			},
+			#[allow(deprecated)]
 			(true, Genesis::Runtime(g)) => {
 				let storage = g.build_storage()?;
-				let top =
-					storage.top.into_iter().map(|(k, v)| (StorageKey(k), StorageData(v))).collect();
-				let children_default = storage
-					.children_default
-					.into_iter()
-					.map(|(sk, child)| {
-						(
-							StorageKey(sk),
-							child
-								.data
-								.into_iter()
-								.map(|(k, v)| (StorageKey(k), StorageData(v)))
-								.collect(),
-						)
-					})
-					.collect();
-
-				Genesis::Raw(RawGenesis { top, children_default })
+				RawGenesis::from(storage)
 			},
-			(_, genesis) => genesis,
+			(true, Genesis::Raw(raw)) => raw,
+
+			(_, genesis) =>
+				return Ok(ChainSpecJsonContainer {
+					client_spec: self.client_spec.clone(),
+					genesis,
+					code: self.code.clone(),
+				}),
 		};
-		Ok(JsonContainer { client_spec: self.client_spec.clone(), genesis })
+
+		if !self.code.is_empty() {
+			raw_genesis.top.insert(
+				StorageKey(sp_core::storage::well_known_keys::CODE.to_vec()),
+				StorageData(self.code.clone()),
+			);
+		}
+
+		Ok(ChainSpecJsonContainer {
+			client_spec: self.client_spec.clone(),
+			genesis: Genesis::Raw(raw_genesis),
+			code: vec![],
+		})
 	}
 
-	/// Dump to json string.
+	/// Dump the chain specification to JSON string.
+	///
+	/// During conversion to `raw` format, the `ChainSpec::code` field will be removed and placed
+	/// into `RawGenesis` as `genesis::top::raw::0x3a636f6465` (which is
+	/// [`sp_core::storage::well_known_keys::CODE`]). If the spec is already in `raw` format, and
+	/// contains `genesis::top::raw::0x3a636f6465` field it will be updated with content of `code`
```

Comment:
You are right. I just wanted to make it super-clear, that it will be overwritten. The alternative scenario/implementation would be to not touch the genesis storage and skip the code field. 

What is intuitive behavior is not obvious, this is why I decided to state it explicitly.